### DISCUSSION TITLE
Which date property does Immich use to sort files in the timeline?

### DISCUSSION THREAD

**Discussion by srahimeen** [immich-app/immich#6459] [Q&A]
I'm seeing some of my photos and videos showing up out of order in the timeline and am trying to figure out why this is happening. I am using the external Library feature to view photos saved from my Google Photos (via Google Takeout).

For example, I have this photo:

```
exiftool 20231126_212635.jpg
ExifTool Version Number         : 12.73
File Name                       : 20231126_212635.jpg
Directory                       : .
File Size                       : 5.1 MB
File Modification Date/Time     : 2023:11:26 21:26:36-08:00
File Access Date/Time           : 2024:01:17 09:30:44-08:00
File Creation Date/Time         : 2024:01:08 20:56:00-08:00
File Permissions                : -rw-rw-rw-
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.01
Exif Byte Order                 : Little-endian (Intel, II)
Make                            : samsung
Camera Model Name               : SM-G970U1
Orientation                     : Horizontal (normal)
X Resolution                    : 72
Y Resolution                    : 72
Resolution Unit                 : inches
Software                        : G970U1UEU9IWH2
Modify Date                     : 2023:11:26 21:26:36
Y Cb Cr Positioning             : Centered
Exposure Time                   : 1/33
F Number                        : 1.9
Exposure Program                : Program AE
ISO                             : 800
Exif Version                    : 0220
Date/Time Original              : 2023:11:26 21:26:36
Create Date                     : 2023:11:26 21:26:36
Offset Time                     : -08:00
Offset Time Original            : -08:00
Shutter Speed Value             : 1/33
Aperture Value                  : 1.9
Exposure Compensation           : 0
Max Aperture Value              : 1.9
Metering Mode                   : Center-weighted average
Flash                           : No Flash
Focal Length                    : 3.3 mm
Color Space                     : sRGB
Exif Image Width                : 3968
Exif Image Height               : 2232
Interoperability Index          : R98 - DCF basic file (sRGB)
Interoperability Version        : 0100
Exposure Mode                   : Auto
White Balance                   : Auto
Digital Zoom Ratio              : 1
Focal Length In 35mm Format     : 25 mm
Scene Capture Type              : Standard
Image Unique ID                 : bc1f409ce440e5de0000000000000000
Compression                     : JPEG (old-style)
Thumbnail Offset                : 864
Thumbnail Length                : 4109
XMP Toolkit                     : XMP Core 5.5.0
Motion Photo                    : 1
Motion Photo Version            : 1
Motion Photo Presentation Timestamp Us: 2904154
Directory Item Mime             : image/jpeg
Directory Item Semantic         : Primary
Directory Item Length           : 0
Directory Item Padding          : 59
Image Width                     : 3968
Image Height                    : 2232
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:2:0 (2 2)
Aperture                        : 1.9
Image Size                      : 3968x2232
Megapixels                      : 8.9
Scale Factor To 35 mm Equivalent: 7.6
Shutter Speed                   : 1/33
Date/Time Original              : 2023:11:26 21:26:36-08:00
Modify Date                     : 2023:11:26 21:26:36-08:00
Thumbnail Image                 : (Binary data 4109 bytes, use -b option to extract)
Circle Of Confusion             : 0.004 mm
Field Of View                   : 71.5 deg
Focal Length                    : 3.3 mm (35 mm equivalent: 25.0 mm)
Hyperfocal Distance             : 1.45 m
Light Value                     : 3.9
```

In the Immich timeline it shows up under Jan 8, 2024. But it was actually shot on November 26, 2023.

Does this mean Immich is using the `File Creation Date/Time` for sorting and not the `Create Date` or `Date/Time Original` for sorting?

**Comment by schuhbacca** [comment]
![Screenshot_20240117-183348~2](https://github.com/immich-app/immich/assets/5739463/e89c7a0e-c031-48bb-9a03-f31ec1378028)

**Reply by HelloMihai** [reply]
which src file is this set in?

**Reply by HelloMihai** [reply]
found it
server/src/services/metadata.service.ts
https://github.com/immich-app/immich/blob/19740a35603156262087c090588c0d9dce89e8ad/server/src/services/metadata.service.ts#L28-L38

**Reply by kiyospace** [reply]
> ![Screenshot_20240117-183348~2](https://private-user-images.githubusercontent.com/5739463/297549376-e89c7a0e-c031-48bb-9a03-f31ec1378028.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ1MzUwOTAsIm5iZiI6MTc1NDUzNDc5MCwicGF0aCI6Ii81NzM5NDYzLzI5NzU0OTM3Ni1lODljN2EwZS1jMDMxLTQ4YmItOWEwMy1mMzFlYzEzNzgwMjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MDdUMDI0NjMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzExNmMwOTlmNDY2NTAxMjQ0ODQzN2IxY2IxOTFhMjQxYjIyMTM0NmFlYjQ4ODQyYmE3YmYxNjk4ZGMzYjEzNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.l3iUHzli1NH8TIyy7Ti0a8jzg4VLZaefTPWcwsMQHVo)

Does this mean that immich would use the 1st valid Date info from this array of exif tags?
The problem is that, in my case, a lot of my photos were sent via SNS messengers. Although they were sent with the "Original File" option checked, when I download photos from SNS messengers, the "SubSecDateTimeOriginal" and "DateTimeOriginal" got changed to my downloading DateTime instead of being kept as the photo's taken DateTime... As I don't download them instantly, sometimes several days later than the time sent, the date info in immich is not accurate...

I would suggest change the order to the following (if changing the order can solve the problem described above) and add 2 more tags (`GPSDateStamp` and `GPSDateTime`). Thanks!

```
 /** look for a date from these tags (in order) */ 
 const EXIF_DATE_TAGS: Array<keyof Tags> = [ 
   'SubSecCreateDate', 
   'CreationDate', 
   'CreateDate',
   'GPSDateStamp',
   'GPSDateTime',
   'SubSecMediaCreateDate', 
   'MediaCreateDate', 
   'DateTimeCreated', 
   'SubSecDateTimeOriginal', 
   'DateTimeOriginal', 
 ]; 
```

Or if the immich Date is not taking the 1st valid date value from the above array as I guessed, could someone propose something to get the REAL Date info for photos sent via SNS messengers? Thanks

**Comment by srahimeen** [comment]
Since in my photo it is `Date/Time Original: 2023:11:26 21:26:36-08:00`, it showing up under Jan 8, 2024 in the timeline is a bug then?

**Comment by bhugh** [comment]
If you use Exiftool you can look at the actual tags, tag groups & families, and tag names vs descriptions.  Immich is using the tag NAME whereas your listing above is showing the DESCRIPTION.

Use the -s option when you run exiftool, to see the actual tag name instead of just the description.  Use options like -G1 -G2 -G3 ... -G9 and -g1 ... -g9 to look at tags in various families and with descriptions vs names.  It's all quite complex and almost certainly when you see the tag names the reason the 2024 date is chosen will become clear.  It's possible, also the Immich is somehow ignoring tags within certain groups or families or whatever for one reason or another. It's also possible that there is another tag, not listed in your output above, that is taking precedence.

Also, it is not clear without digging into the code further, what happens if there are two tags with the same name - as can very frequently happen.  So if there are two DateTime Original tags, is the first one or last one encountered, or the earliest in date, or latest in date, or what?  Something like that could easily be in play.

https://exiftool.org

Decent explanation of the different groups, families, etc: https://adamtheautomator.com/exiftool/

Explanation of groups vs families: https://exiftool.org/forum/index.php?topic=11326.0  and https://exiftool.org/#groups

List of some 27000 known tags used in various files, devices, and file formats: https://exiftool.org/TagNames/index.html

**Comment by bhugh** [comment]
Also if wanting to definitely set the date to be used, it looks like the simplest way might be to use a tool like Exiftool to set `SubSecDateTimeOriginal`.

Update: After poking around, it looks like SubSecDateTimeOriginal and similar SubSec... date/times are not actually set directly.  Rather Exif has a field for DateTimeOriginal, and also (in some cases!) for SubSecTimeOriginal, and also OffsetTimeOriginal.  So here is a typical example:

```ModifyDate                      : 2022:05:16 17:23:26
DateTimeOriginal                : 2022:05:16 17:23:25
CreateDate                      : 2022:05:16 17:23:26
OffsetTime                      : -05:00
OffsetTimeOriginal              : -05:00
OffsetTimeDigitized             : -05:00
SubSecTime                      : 132321
SubSecTimeOriginal              : 132321
SubSecTimeDigitized             : 132321
GPSTimeStamp                    : 22:23:19
GPSDateStamp                    : 2022:05:16
ProfileDateTime                 : 2016:12:08 09:38:28
SubSecCreateDate                : 2022:05:16 17:23:26.132321-05:00
SubSecDateTimeOriginal          : 2022:05:16 17:23:25.132321-05:00
SubSecModifyDate                : 2022:05:16 17:23:26.132321-05:00
```

That is the output of 'exiftool -s -time:all myimage.jpg'.  I _believe_ this is similar to the level of detail Immich looks at when deciding which is the "real" date of the file.

Here is the output from 'exiftool -s -g1 -time:all myimage.jpg', which gives more detail into the structure of the exif data:

```
---- System ----
FileModifyDate                  : 2022:05:16 17:23:25-05:00
FileAccessDate                  : 2024:01:31 23:40:09-06:00
FileCreateDate                  : 2022:05:16 17:23:25-05:00
---- IFD0 ----
ModifyDate                      : 2022:05:16 17:23:26
---- ExifIFD ----
DateTimeOriginal                : 2022:05:16 17:23:25
CreateDate                      : 2022:05:16 17:23:26
OffsetTime                      : -05:00
OffsetTimeOriginal              : -05:00
OffsetTimeDigitized             : -05:00
SubSecTime                      : 132321
SubSecTimeOriginal              : 132321
SubSecTimeDigitized             : 132321
---- GPS ----
GPSTimeStamp                    : 22:23:19
GPSDateStamp                    : 2022:05:16
---- ICC-header ----
ProfileDateTime                 : 2016:12:08 09:38:28
---- Composite ----
SubSecCreateDate                : 2022:05:16 17:23:26.132321-05:00
SubSecDateTimeOriginal          : 2022:05:16 17:23:25.132321-05:00
SubSecModifyDate                : 2022:05:16 17:23:26.132321-05:00
GPSDateTime                     : 2022:05:16 22:23:19Z
```

Point being, DateTimeOriginal, OffsetTimeOriginal, and SubSecTimeOriginal are 3 separate fields in the ExifIFD data.  They are then combined to make the composite field SubSecDateTimeOriginal, which is the field that Immich is going to look at.

Similarly for SubSecCreateDate and SubSecModifyDate.

Just for example, with exiftool, you could do this:

`exiftool "-Exif:DateTimeOriginal=1912:05:15 2:23:26" "myimage.jpg"`

That would change only the ExifIFD:DateTimeOriginal field, leaving SubSecTimeOriginal  and OffsetTimeOriginal unchanged.  The compose SubSecTimeOriginal would have the new DateTime with the original subseconds and OffsetTime.

On the other hand, you can do this:

`exiftool "-DateTimeOriginal=1912:05:15 2:23:26" "myimage.jpg"`

which overwrites the composite DateTimeOriginal - meaning that the OffsetTime and SubSeconds are now gone.

A third option is:

`exiftool "-SubSecDateTimeOriginal=1912:05:15 2:23:26.321123-05:00" "myimage.jpg"`

This will fill in ExifIFD:DateTimeOriginal=1012:05:15, ExifIFD:SubSecTimeOriginal=321123, and ExifIFD:OffsetTimeOriginal=-05:00.

Equivalent to this is setting each of those fields specifically:

`exiftool "-Exif:DateTimeOriginal=1912:05:15 2:23:26" "-Exif:SubSecTimeOriginal=321123" "-Exif:OffsetTimeOriginal=-05:00" "myimage.jpg"`

There is some discussion of this here: https://exiftool.org/forum/index.php?topic=13143.0

**Comment by aksnet** [comment]
Hi,
i found this entry cause my photo overview is listing all my converted movies as first entry, also some photos which have no exif entries are not listed correct.
I am using only the external library function to show my photos. I do not upload any of them via the immich page.

All my photos and videos are named this way: 2017-01-06 16-40-00 Andi.JPG/heiv/mp4/mov and so on.
Is there a way to tell immich sort the photos by using the filename <year>-<month>-<day> <hour>-<minute>-<second>(and then ignore all behind).<extension>

Thanks
Andi

**Reply by axelkar** [reply]
I think what you should do is take the time and date information from the filename and put it into the EXIF data using exiftool. Then it'll work with your file manager or when sending/downloading the file by itself.

https://rb.ax/blog/using-exiftool-to-set-dates-based-on-filename/
https://exiftool.org/forum/index.php?topic=13064.0

**Comment by axelkar** [comment]
It'd be nice if Immich showed me which property it was using and the other properties not normally shown, and warned when an asset only has the filesystem mtime, and no internal metadata.

**Comment by Kusnierz-Maciej** [comment]
This is time data for my image. And Immich uses the File Modify date for some reason.
```
---- File ----
FileModifyDate                  : 2025:10:21 21:01:49+02:00
FileAccessDate                  : 2025:11:11 12:04:49+01:00
FileCreateDate                  : 2021:01:30 18:28:12+01:00
---- EXIF ----
ModifyDate                      : 2007:06:02 14:25:25
DateTimeOriginal                : 2007/04/29 08:19:45
CreateDate                      : 2007/04/29 08:19:45
---- XMP ----
MetadataDate                    : 2007:06:02 14:25:25+02:00
DateTimeDigitized               : 2007/04/29 08:19:45
---- ICC_Profile ----
ProfileDateTime                 : 1998:02:09 06:49:00
```

**Reply by sbuckmann** [reply]
EXIF date structure is YYYY:MM:DD, not YYYY/MM/DD.

**Reply by Kusnierz-Maciej** [reply]
That was the issue. Wrote script to fix dates and it works prefect now.

**Comment by NateMarmoton** [comment]
Seems like a simple, possibly more robust, solution would be to use the oldest date related field attached to an image, rather then prioritizing specific field names over another arbitrarily.
